Updated models for reproducibility

dongbohu opened the pull request:
## Issue Number

#320 
#322
#324 

## Purpose/Implementation Notes

Updated models for the two issues above

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests

List out the functional tests you've completed to verify your changes work locally.

## Checklist

_Put an `x` in the boxes that apply._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots

Please attach any screenshots that illustrate these changes.

dongbohu commented:
@kurtwheeler and @Miserlou: This PR includes the updated models that we discussed on Monday. Since the field in `ComputationResult` is renamed from `command_executed` to `commands` and the removal of `program_version` and `system_version`, the unit tests will fail. Before I fix the lines that are affected by the new models, I want to make sure that this is what we were talking about. Thanks.

dongbohu commented on common/data_refinery_common/models/models.py:
```diff
@@ -366,6 +364,29 @@ def save(self, *args, **kwargs):
+    "Pipeline that is associated with a series of ComputationalResult records."""
+
+    name = models.CharField(max_length=255)
+    steps = ArrayField(models.IntegerField(), default=[])
```
Ideally we would want to have something like this instead:
```
steps = ArrayField(
        models.ForeignKey(ComputationalResult, on_delete=models.CASCADE),
        default=[]
) 
```
But it is not allowed in Django, see:
https://docs.djangoproject.com/en/2.0/ref/contrib/postgres/fields/#django.contrib.postgres.fields.ArrayField.base_field

So I am using the regular integer to refer to the primary key of related ComputationalResult record.

dongbohu commented on common/data_refinery_common/models/models.py:
```diff
@@ -306,13 +306,11 @@ def __str__(self):
-        max_length=255)  # Generally defined in from data_refinery_workers._version import __version__
     is_ccdl = models.BooleanField(default=True)
 
+    # TODO: This field should be changed into "processor" (foreign key to "Processor").
```
This is a reminder to ourselves for further changes of the model to fix #324.

kurtwheeler commented:
This looks correct to me.

Miserlou commented on common/data_refinery_common/models/models.py:
```diff
@@ -306,13 +306,11 @@ def __str__(self):
     # Managers
     objects = models.Manager()
     public_objects = PublicObjectsManager()
+    commands = ArrayField(models.TextField())
```
This should be spatially separated from the Manger objects here, which perform a different function

Miserlou commented on common/data_refinery_common/models/models.py:
```diff
@@ -366,6 +364,29 @@ def save(self, *args, **kwargs):
+    """Processor associated with a certain ComputationalResult."""
+
+    name = models.CharField(max_length=255)
+    docker_img = models.CharField(max_length=255)
```
We should use pythonic rather than C-like names here, so `docker_image` and `environment`, please.

It'd also be great if these could have a `__str__` method so that they display nicely in the Django admin interface.
Looks good otherwise!

dongbohu replied:
Thank you @Miserlou!

dongbohu commented:
I think this PR is ready for review. I have removed "Work in progress" for its title. It also addressed isse #324 partially. With the new `processor` field added in `ComputationalResult` as a foreign key, `pipeline` field in `ComputationalResult` becomes redundant. I tried to remove it but then realized that it will break some other code that @Miserlou is working on. (As you can see, this PR already conflicts with a few files in `dev` branch.) So I am leaving both fields in for now. When @Miserlou is done with his work, one of us needs to file another PR to resolve the redundancy.

I didn't include the YAML config files (and `Processor` table population based on them) here, because:
(1) This PR is already big enough;
(2) I need more time to finalize the YML files. 

I am taking next week off, but I will be accessible on the Internet. So please feel free to add your comments.

Miserlou commented on common/data_refinery_common/models/models.py:
```diff
@@ -294,6 +294,30 @@ def save(self, *args, **kwargs):
+class Pipeline(models.Model):
+    "Pipeline that is associated with a series of ComputationalResult records."""
+
+    name = models.CharField(max_length=255)
```
After review the code, I'm thinking that this should maybe be a ChoiceField or at least use an Enum or something somwhere so we don't have the problem of having to know `No Op` versus `noop` versus `NO_OP` vs etc. etc. etc. 

This goes for everything

dongbohu replied:
I can definitely see the advantages of using Enum or ChoiceField. There is one disadvantage though: Every time when we add a new pipeline, we will have to modify the model and migrate the database.

kurtwheeler replied:
Well we don't necessarily need to have it be at the database level. I think we need it somewhere, it could be an Enum class or even just a namespace full of `SALMON_QUANT = "SALMON_QUANT"` so that you could do `PipelineEnums.SALMON_QUANT` instead of hardcoding the string. I think this might help looking up what the possible values are along with keeping them consistent. We have at least the following values:
`"Salmon Quant", "Tximport", "MultiQC", "Salmontools", "Submitter-processed", "Agilent SCAN TwoColor"` which aren't very consistent case-wise.

I think I would prefer a naming convention like `ALL_CAPS_v1.2.3`. 

However I was also thinking that it might make sense to actually have the version be a separate field so that if we ever wanted to get all the pipelines for `AGILENT_SCAN_TWOCOLOR` we could just query with `WHERE name='AGILENT_SCAN_TWOCOLOR' rather than `WHERE name LIKE 'AGILENT_SCAN_TWOCOLOR%'`. Also since we're using the version of the system for the version number we could query for all results from a given version.

dongbohu replied:
In the new database model, the terms such as "Salmon Quant", "Tximport", "MultiQC", "Salmontools", "Submitter-processed" and "Agilent SCAN TwoColor" are actually names of **processors**, not names of **pipelines**. 

The possible values of pipelines are hardcoded at the end of each pipeline module (`salmon.py`, `agilent_twocolor.py`, etc)  right before the call of `utils.run_pipeline()`. 

The naming convention of `ALL_CAPS_v1.2.3` is a good point. But I'm not sure whether these processor names will be shown on web UIs. If they will, then a name with white spaces seems to be more friendly to the front end because it can be displayed directly.  

I agree that having the processor and pipelines hardcoded everywhere is not a good idea. Let me put them at the beginning of each pipeline module.

Having a separate `version` field is a good idea too. I will add it in the model. Thanks.

kurtwheeler replied:
Ah yeah that's a good point about user-friendly names.

kurtwheeler commented on workers/data_refinery_workers/processors/array_express.py:
```diff
@@ -144,19 +153,17 @@ def _create_result_objects(job_context: Dict) -> Dict:
 
     result = ComputationalResult()
-    result.command_executed = "'SCAN.UPC', 'SCANfast'" # Need a better way to represent this R code.
+    result.commands = ['SCAN.UPC', 'SCANfast'] # Need a better way to represent this R code.
```
This should be `"SCAN.UPC::SCANfast"` because this is the same as `"SCAN.UPC::SCAN_TwoColor"`, just a different function from the SCAN.UPC package.

kurtwheeler commented on workers/data_refinery_workers/processors/no_op.py:
```diff
@@ -34,11 +42,13 @@ def _no_op_processor_fn(job_context: Dict) -> Dict:
     # This is a NO-OP, but we make a ComputationalResult regardless.
     result = ComputationalResult()
-    result.command_executed = "" # No op!
+    result.commands.append("") # No op!
```
I think this should actually have a command name, especially since once Rich's recent work goes out the NO_OP command is actually going to be more of a CONVERT_GENE_IDS command.

dongbohu replied:
Are you saying that this line should be changed to:
```
result.commands.append("CONVERT_GENE_IDS")
```

kurtwheeler replied:
Maybe? At the moment all the gene conversion work is still using NO_OP as the name, but it's becoming a misnomer because it's actually doing a decent amount of work.

Let's keep it as NO_OP for now, and then we can think about using a CONVERT_GENE_IDS name in a broader sense at some point.

dongbohu commented on workers/data_refinery_workers/processors/array_express.py:
```diff
@@ -153,15 +153,15 @@ def _create_result_objects(job_context: Dict) -> Dict:
 
     result = ComputationalResult()
-    result.commands = ['SCAN.UPC', 'SCANfast'] # Need a better way to represent this R code.
+    result.commands.append('SCAN.UPC::SCANfast') # Need a better way to represent this R code.
```
Agree. I have removed this comment. If there's nothing else, I will merge this PR. Thanks.

kurtwheeler replied:
Do it!
